RDoc-2642 [Python] Client API > Session > Querying > Query overview [Replace C# samples] (#1778)
ravendb/docs: Documentation/5.4/Raven.Documentation.Pages/client-api/session/querying/how-to-query.python.markdown

# Query Overview

---

{NOTE: }

* Queries in RavenDB can be written with either of the following:

>     * __LINQ__ - when querying with the session's `query` method.  

Reviewer: there's no LINQ in Python 
```suggestion
    The session's `query` method - rich API is provided
```

Author: done

  Entities will Not be tracked when:  
    * Query returns a [projection](../../../client-api/session/querying/how-to-project-query-results)  
    * Tracking is [disabled](../../../client-api/session/configuration/how-to-disable-tracking#disable-tracking-query-results)  

* Query results are __cached__ by default. To disable query caching see [NoCaching](../../../client-api/session/querying/how-to-customize-query#nocaching).

* Queries are timed out after a configurable time period.  See [query timeout](../../../server/configuration/database-configuration#databases.querytimeoutinsec).

---

* In this page:
  * [Queries always provide results using an index](../../../client-api/session/querying/how-to-query#queries-always-provide-results-using-an-index)  
  * [session.query](../../../client-api/session/querying/how-to-query#session.query)  
  * [session.advanced.document_query](../../../client-api/session/querying/how-to-query#session.advanced.document_query)  
  * [session.advanced.raw_query](../../../client-api/session/querying/how-to-query#session.advanced.raw_query)

>   * [Custom methods and extensions for LINQ](../../../client-api/session/querying/how-to-query#custom-methods-and-extensions-for-linq)  

Reviewer: no linq in python

{NOTE/}

{NOTE: }

<a id="collectionQuery" /> 
__3. Query a collection - query full collection | query by ID__:  
 
* __Query type__: Full collection Query  
  __Index used__: The raw collection (internal storage indexes)  

* Full collection query:

  * When querying a collection without specifying an index and with no filtering condition,  
    then all documents from the specified collection are returned.

  * RavenDB uses the raw collection documents in its __internal storage indexes__ as the source for this query.  
    No auto-index is created.
   
  * Example RQL: &nbsp; `from Employees`

* Query by document ID:
 
  * When querying a collection only by document ID or IDs,  
    then similar to the full collection query, no auto-index is created.  
  
  * RavenDB uses the raw collection documents as the source for this query.  

  * Example RQL: &nbsp; `from Employees where id() == "employees/1-A"`  
    See more examples [below](../../../client-api/session/querying/how-to-query#session.query).

{NOTE/}

{PANEL/}

{PANEL: session.query}

> * The simplest way to issue a query is by using the session's `query` method which supports LINQ.  

Reviewer: no linq here

Author: removed

>   Both the LINQ method syntax and the LINQ query syntax are supported.  

Reviewer: ```suggestion
 
```

Author: removed

from "Employees" where id() == "employees/1-A"
{CODE-TAB-BLOCK/}
{CODE-TABS/}

{NOTE/}

{NOTE: }

__Query collection - with filtering__ 

{CODE-TABS}
{CODE-TAB:python:Method-syntax query_3_1@ClientApi\Session\Querying\HowToQuery.py /}
{CODE-TAB:python:Query-syntax query_3_3@ClientApi\Session\Querying\HowToQuery.py /}
{CODE-TAB-BLOCK:sql:RQL}
// Query collection - filter by document field

// An auto-index will be created if there isn't already an existing auto-index
// that indexes the requested field

from "Employees" where FirstName == "Robert"
{CODE-TAB-BLOCK/}
{CODE-TABS/}

{NOTE/}

{NOTE: }

__Query collection - with paging__ 

{CODE-TABS}
{CODE-TAB:python:Method-syntax query_4_1@ClientApi\Session\Querying\HowToQuery.py /}
{CODE-TAB:python:Query-syntax query_4_3@ClientApi\Session\Querying\HowToQuery.py /}
{CODE-TAB-BLOCK:sql:RQL}
// Query collection - page results
// No auto-index is created since no filtering is applied

from "Products" limit 5, 10 // skip 5, take 10
{CODE-TAB-BLOCK/}
{CODE-TABS/}

* By default, if the page size is not specified, all matching records will be retrieved from the database.

{NOTE/}

{PANEL/}

{PANEL: session.advanced.document_query}

> * `document_query` provides a full spectrum of low-level querying capabilities,  

Reviewer: it's not true - simple session.query gives also full capabilities, we have document_query in python but it ain't something amazing like in C#, it has the same capabilities like the session.query()

Author: done

    * [query -vs- document_query](../../../client-api/session/querying/document-query/query-vs-document-query)

__Example__:

{CODE-TABS}
{CODE-TAB:python:document_query query_5_1@ClientApi\Session\Querying\HowToQuery.py /}
{CODE-TAB-BLOCK:sql:RQL}
// Query collection - filter by document field

// An auto-index will be created if there isn't already an existing auto-index
// that indexes the requested field

from "Employees" where FirstName = "Robert"
{CODE-TAB-BLOCK/}
{CODE-TABS/}

{PANEL/}

{PANEL: session.advanced.raw_query}

* Queries defined with [query](../../../client-api/session/querying/how-to-query#session.query) or [document_query](../../../client-api/session/querying/how-to-query#session.advanced.document_query) are translated by the RavenDB client to [RQL](../../../client-api/session/querying/what-is-rql)  
  when sent to the server.

* The session also gives you a way to express the query directly in RQL using the `raw_query` method.

__Example__:

{CODE-TABS}
{CODE-TAB:python:RawQuery query_6_1@ClientApi\Session\Querying\HowToQuery.py /}
{CODE-TABS/}

{PANEL/}

> {PANEL: Custom methods and extensions for LINQ}

Reviewer: no linq here, but i can recognize these methods. I've implemented most of them

I haven't implemented yet:
- `group_by_array_content`
- `group_by_array_values`

Author: - changed the title and removed the two unimplemented methods  
- please remember to use the "python documentation required" label when you implement them so their description can be readded here

- [long_count](../../../client-api/session/querying/how-to-count-query-results)
- [more_like_this](../../../client-api/session/querying/how-to-use-morelikethis)
- [if_type](../../../client-api/session/querying/how-to-project-query-results#oftype-(as)---simple-projection)
- [order_by_distance](../../../client-api/session/querying/how-to-make-a-spatial-query#orderByDistance)
- [order_by_distance_descending](../../../client-api/session/querying/how-to-make-a-spatial-query#orderByDistanceDesc)
- [order_by_score](../../../client-api/session/querying/sort-query-results#order-by-score)
- [order_by_score_descending](../../../client-api/session/querying/sort-query-results#order-by-score)
- [project_into](../../../client-api/session/querying/how-to-project-query-results)
- [search](../../../client-api/session/querying/text-search/full-text-search)
- [spatial](../../../client-api/session/querying/how-to-make-a-spatial-query)
- [statistics](../../../client-api/session/querying/how-to-get-query-statistics)
- [suggest_using](../../../client-api/session/querying/how-to-work-with-suggestions)

{PANEL/}

{PANEL: Syntax}

{CODE:python syntax@ClientApi\Session\Querying\HowToQuery.py /}

> | Parameter | Type | Description |

Reviewer: this description don't match methods.
```py
 def query(
                self, source: Optional[Query] = None, object_type: Optional[Type[_T]] = None
            ) -> DocumentQuery[_T]:
                ...

            def query_collection(
                self, collection_name: str, object_type: Optional[Type[_T]] = None
            ) -> DocumentQuery[_T]:
                ...

            def query_index(self, index_name: str, object_type: Optional[Type[_T]] = None) -> DocumentQuery[_T]:
                ...

            def document_query(
                self,
                index_name: str = None,
                collection_name: str = None,
                object_type: Type[_T] = None,
                is_map_reduce: bool = False,
            ) -> DocumentQuery[_T]:
                ...

            # Overloads for querying an index:
            # ================================
            def query_index_type(
                self, index_type: Type[_TIndex], object_type: Optional[Type[_T]] = None
            ) -> DocumentQuery[_T]:
                ...

            def document_query_from_index_type(
                self, index_type: Type[_TIndex], object_type: Type[_T]
            ) -> DocumentQuery[_T]:
                ...

            # RawQuery
            # ================================
            def raw_query(self, query: str, object_type: Optional[Type[_T]] = None) -> RawDocumentQuery[_T]:
                ...
```

Reviewer: i thought we're going to describe each method, and as python lacks overrides we'd need to describe each `method-with-different-name(override)`
